File: src/exoticatechnologies/modifications/upgrades/UpgradesHandler.java
package exoticatechnologies.modifications.upgrades;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import data.scripts.util.MagicSettings;
import exoticatechnologies.ui.impl.shop.ShopManager;
import exoticatechnologies.ui.impl.shop.exotics.ExoticShopUIPlugin;
import exoticatechnologies.ui.impl.shop.upgrades.UpgradeShopUIPlugin;
import exoticatechnologies.ui.impl.shop.upgrades.methods.*;
import lombok.extern.log4j.Log4j;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;
import java.util.*;

@Log4j
public class UpgradesHandler {
    public static final Map<String, Upgrade> UPGRADES = new HashMap<>();
    public static final List<Upgrade> UPGRADES_LIST = new ArrayList<>();
    public static final Set<UpgradeMethod> UPGRADE_METHODS = new LinkedHashSet<>();

    public static void addUpgradeMethod(UpgradeMethod method) {
        UPGRADE_METHODS.add(method);
    }

    public static void initialize() {
        UPGRADES.clear();
        UPGRADE_METHODS.clear();
        UPGRADE_METHODS.add(new CreditsMethod());
        UPGRADE_METHODS.add(new ResourcesMethod());
        UPGRADE_METHODS.add(new ChipMethod());
        UPGRADE_METHODS.add(new RecoverMethod());


        UpgradesHandler.populateUpgrades();

        UPGRADES_LIST.clear();
        List<String> orderedKeys = MagicSettings.getList("exoticatechnologies", "upgradeOrder");
        for (int i = 0; i < orderedKeys.size(); i++) {
            String key = orderedKeys.get(i);
            UPGRADES_LIST.add(UPGRADES.get(key));
        }

        for (Upgrade upgrade : UPGRADES.values()) {
            if (!orderedKeys.contains(upgrade.getKey())) {
                UPGRADES_LIST.add(upgrade);
            }
        }

        ShopManager.addMenu(new UpgradeShopUIPlugin());
    }

    public static void populateUpgrades() {
        try {
            JSONObject settings = Global.getSettings().getMergedJSONForMod("data/config/upgrades.json", "exoticatechnologies");

            Iterator upgIterator = settings.keys();
            while (upgIterator.hasNext()) {
                String upgKey = (String) upgIterator.next();

                if (UPGRADES.containsKey(upgKey)) continue;

                Upgrade upgrade;
                try {
                    JSONObject upgradeSettings = settings.getJSONObject(upgKey);

                    if (upgradeSettings.has("upgradeClass")) {
                        Class<?> clzz = Global.getSettings().getScriptClassLoader().loadClass(upgradeSettings.getString("upgradeClass"));

                        //magic to get around reflection block
                        upgrade = (Upgrade) MethodHandles.lookup().findConstructor(clzz, MethodType.methodType(void.class, String.class, JSONObject.class))
                                .invoke(upgKey, upgradeSettings);
                        if (!upgrade.shouldLoad()) {
                            upgrade = null;
                        }
                    } else {
                        upgrade = new Upgrade(upgKey, upgradeSettings);

                        if (!upgrade.shouldLoad()) {
                            upgrade = null;
                        }
                    }

                    if (upgrade != null) {
                        UpgradesHandler.addUpgrade(upgrade);

                        log.info(String.format("loaded upgrade [%s]", upgrade.getName()));
                    }
                } catch (JSONException ex) {
                    String logStr = String.format("Upgrade [%s] had an error.", upgKey);
                    log.error(logStr);
                    throw new RuntimeException(logStr, ex);
                }
            }
        } catch (Throwable ex) {
            throw new RuntimeException(ex);
        }
    }

    public static void addUpgrade(Upgrade upgrade) {
        if (UPGRADES.containsKey(upgrade.getKey())) return;

        UPGRADES.put(upgrade.getKey(), upgrade);
    }
}

File: src/exoticatechnologies/modifications/upgrades/UpgradesHandler.kt
package exoticatechnologies.modifications.upgrades

import com.fs.starfarer.api.Global
import exoticatechnologies.ui.impl.shop.ShopManager.Companion.addMenu
import exoticatechnologies.ui.impl.shop.upgrades.UpgradeShopUIPlugin
import exoticatechnologies.ui.impl.shop.upgrades.methods.UpgradeMethod
import lombok.extern.log4j.Log4j
import org.apache.log4j.Logger
import org.json.JSONException
import org.json.JSONObject
import org.magiclib.util.MagicSettings
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType

@Log4j
object UpgradesHandler {
    private val log = Logger.getLogger(UpgradesHandler.javaClass)

    @JvmField
    val UPGRADES: MutableMap<String, Upgrade> = HashMap()
    @JvmField
    val UPGRADES_LIST: MutableList<Upgrade> = ArrayList()
    @JvmField
    val UPGRADE_METHODS: MutableSet<UpgradeMethod> = LinkedHashSet()
    @JvmField
    val UPGRADES_BY_HINT: MutableMap<String, MutableSet<Upgrade>> = LinkedHashMap()

    fun addUpgradeMethod(method: UpgradeMethod) {
        UPGRADE_METHODS.add(method)
    }

    @JvmStatic
    fun initialize() {
        populateUpgrades()
        populateMethods()
        addMenu(UpgradeShopUIPlugin())
    }

    fun populateUpgrades() {
        UPGRADES.clear()

        val settings = Global.getSettings().getMergedJSONForMod("data/config/upgrades.json", "exoticatechnologies")
        val upgIterator = settings.keys()
        while (upgIterator.hasNext()) {
            val upgKey = upgIterator.next() as String
            if (UPGRADES.containsKey(upgKey)) continue
            var upgrade: Upgrade?
            try {
                val upgradeSettings = settings.getJSONObject(upgKey)
                if (upgradeSettings.has("upgradeClass")) {
                    val clzz =
                        Global.getSettings().scriptClassLoader.loadClass(upgradeSettings.getString("upgradeClass"))

                    //magic to get around reflection block
                    upgrade = MethodHandles.lookup().findConstructor(
                        clzz,
                        MethodType.methodType(Void.TYPE, String::class.java, JSONObject::class.java)
                    )
                        .invoke(upgKey, upgradeSettings) as Upgrade
                    if (!upgrade.shouldLoad()) {
                        upgrade = null
                    }
                } else {
                    upgrade = Upgrade(upgKey, upgradeSettings)
                    if (!upgrade.shouldLoad()) {
                        upgrade = null
                    }
                }
                if (upgrade != null) {
                    addUpgrade(upgrade)
                    log.info(String.format("loaded upgrade [%s]", upgrade.name))
                }
            } catch (ex: JSONException) {
                val logStr = String.format("Upgrade [%s] had an error.", upgKey)
                log.error(logStr)
                throw RuntimeException(logStr, ex)
            }
        }

        UPGRADES_LIST.clear()
        val orderedKeys = MagicSettings.getList("exoticatechnologies", "upgradeOrder")
        for (i in orderedKeys.indices) {
            val key = orderedKeys[i]
            UPGRADES[key]?.let {
                UPGRADES_LIST.add(it)
            }
        }
        for (upgrade in UPGRADES.values) {
            if (!orderedKeys.contains(upgrade.key)) {
                UPGRADES_LIST.add(upgrade)
            }
        }
    }

    fun populateMethods() {
        UPGRADE_METHODS.clear()
        try {
            val settings = Global.getSettings().getMergedSpreadsheetDataForMod("name", "data/config/upgradeMethods.csv", "exoticatechnologies")
            for (i in 0 until settings.length()) {
                val methodData = settings.getJSONObject(i)
                val methodClassPath = methodData.getString("class")
                val clzz =
                    Global.getSettings().scriptClassLoader.loadClass(methodClassPath)
                val method = MethodHandles.lookup().findConstructor(
                    clzz,
                    MethodType.methodType(Void.TYPE)
                )
                    .invoke() as UpgradeMethod

                if (method.shouldLoad()) {
                    UPGRADE_METHODS.add(method)
                }
            }
        } catch (ex: Throwable) {
            throw RuntimeException(ex)
        }
    }

    fun addUpgrade(upgrade: Upgrade) {
        if (UPGRADES.containsKey(upgrade.key)) return
        UPGRADES[upgrade.key] = upgrade

        upgrade.hints.forEach {
            UPGRADES_BY_HINT.getOrPut(it) { LinkedHashSet() }
                .add(upgrade)
        }
    }
}